R/derive_locf_records.R

```diff
@@ -0,0 +1,188 @@
+#'   "CDISC01", "01-701-1028", "SYSBP", 130, 0, "BASELINE",
+#'   "CDISC01", "01-701-1028", "SYSBP", 132, 2, "WEEK 2"
+#' )
+#'
+#'
+#' # A dataset with all the combinations of PARAMCD, PARAM, AVISIT, AVISITN, ... which are expected.
+#' advs_expected_obsv <- tibble::tribble(
+#'   ~PARAMCD, ~AVISITN, ~AVISIT,
+#'   "PULSE", 0, "BASELINE",
+#'   "PULSE", 0, "WEEK 6",
+#'   "DIABP", 0, "BASELINE",
+#'   "DIABP", 2, "WEEK 2",
+#'   "DIABP", 4, "WEEK 4",
+#'   "DIABP", 6, "WEEK 6",
+#'   "SYSBP", 0, "BASELINE",
+#'   "SYSBP", 2, "WEEK 2",
+#'   "SYSBP", 4, "WEEK 4",
+#'   "SYSBP", 6, "WEEK 6"
+#' )
+#'
+#' derive_locf_records(
+#'   data = advs,
+#'   dataset_expected_obs = advs_expected_obsv,
+#'   by_vars = vars(STUDYID, USUBJID, PARAMCD),
+#'   order = vars(STUDYID, USUBJID, PARAMCD, AVISITN, AVISIT)
+#' )
+#'
+derive_locf_records <- function(dataset,
+                                dataset_expected_obs,
+                                by_vars,
+                                order) {
+  #### Input Checking ####
+
+  # Check if input parameters is a valid list of variables
+  assert_vars(by_vars, optional = TRUE)
+  assert_order_vars(order)
+
+  # Check by_vars and order variables in input datasets
+  assert_data_frame(dataset_expected_obs)
+  assert_data_frame(
+    dataset,
+    required_vars = quo_c(by_vars, extract_vars(order), chr2vars(colnames(dataset_expected_obs)))
+  )
+
+  #### Prepping 'dataset_expected_obs' ####
+
+
+  # Get the STUDYID and USUBJID from input dataset to set expected
+  # observations for each subject
+  subjids <- dataset %>%
+    select(STUDYID, USUBJID) %>%
+    distinct()
+
+  exp_obsv <- subjids %>%
+    crossing(dataset_expected_obs)
+
+
+
+  ##### Add LOCF records ####
+
+  # Split input dataset into the missing and non-missing AVAL records
+  aval_missing <- dataset %>%
+    filter(is.na(AVAL))
+
+  aval_not_missing <- dataset %>%
+    drop_na(AVAL)
+
+
+  # Get the variable names from the expected observation dataset
+  exp_obs_vars <- exp_obsv %>%
+    colnames()
+
+
+  # Get unique combination of visits/timepoints per parameter per subject
+  # from the original input dataset (with non-missing AVAL)
+  advs_unique_original <- aval_not_missing %>%
+    select(all_of(exp_obs_vars)) %>%
+    distinct()
+
+
+  tmp_dtype <- get_new_tmp_var(exp_obsv, prefix = "tmp_dtype")
+
+  # Get all the expected observations that are to be added to the input
+  # dataset (with non-missing AVAL)
+  advs_exp_obsv3 <- exp_obsv %>%
+    mutate(!!tmp_dtype := "LOCF") %>%
+    anti_join(advs_unique_original, by = c(exp_obs_vars)) %>%
+    drop_na(exp_obs_vars)
+
+
+  # Merge the expected observations with the input dataset (with non-missing AVAL)
+  # Arrange the dataset by 'order' and group it by 'by_vars'
+  # Use fill() to fill the AVAL from the previous observation for the newly added records
+
+  attach(aval_not_missing)
+
+  if (exists("DTYPE") == FALSE) {
+    aval_not_missing_locf <- aval_not_missing %>%
```

Comment:
Please break up the chain to avoid repetition of code. E.g.,
```
  aval_not_missing_locf <- aval_not_missing %>%
      full_join(advs_exp_obsv3, by = c(exp_obs_vars))
  if ("DTYPE" %in% colnames(aval_not_missing) {
    aval_not_missing_locf <- aval_not_missing_locf %>%
      mutate(DTYPE = if_else(!!tmp_dtype == "LOCF", "LOCF", DTYPE, missing = DTYPE))) %>%
      select(-!!tmp_dtype)
  } else {
    aval_not_missing_locf <- rename(aval_not_missing_locf, DTYPE = !!tmp_dtype)
  }
  aval_not_missing_locf <- aval_not_missing_locf %>%
    arrange(!!!byvars, !!!order) %>%
    group_by(!!!by_vars) %>%
    fill("AVAL") %>%
    ungroup()
```